### PR TITLE
[Feature] Support Segmenter

### PR DESCRIPTION
**Pull request by h1063135843**
## Motivation

OpenMMLAB competition

## Modification

1.Move vit transform encode layer to utils in order to be reused by segmenter
2.Add Segmenter implementation and configs

## Use cases (Optional)

python tools\train.py configs\segmenter\segmenterL_vit-b16_ln_512x512_160k_ade20k.py

## Checklist
None

### CODE REVIEW

**Comment by Junjun2016** [comment]
Hi @h1063135843 
Please fix the lint error!

**Comment by h1063135843** [comment]
> Hi @h1063135843
> Please fix the lint error!

I have no idea about how to fix it. It is my first time to commit codes. Do I need to add some unit test?

**Comment by rstrudel** [comment]
Hi there!
I am one of the main author of Segmenter. Thanks a lot for opening a PR to integrate Segmenter in mmsegmentation!
How are things going? Is there a way I can help with the PR?

**Comment by MengzhangLI** [comment]
> Hi there! I am one of the main author of Segmenter. Thanks a lot for opening a PR to integrate Segmenter in mmsegmentation! How are things going? Is there a way I can help with the PR?

Hi of course!

This PR is created by community and it is not maintained, so I think you could create a new PR if you think it is more convenient.

To keep uniform format of other models, could you please follow other transformer models such as [segformer](https://github.com/open-mmlab/mmsegmentation/tree/master/configs/segformer) and [swin](https://github.com/open-mmlab/mmsegmentation/tree/master/configs/swin). I think refactory code is necessary thus we would review your code once it was pulled requested.

Thanks again for your nice help! Let's work together to make more people use `segmenter`.

Best,

**Comment by h1063135843** [comment]
> Hi there! I am one of the main author of Segmenter. Thanks a lot for opening a PR to integrate Segmenter in mmsegmentation! How are things going? Is there a way I can help with the PR?

Thanks for your attention.
I have only a single gpu 1080Ti, which doesn't support me to reproduce comparable accuracy as your article.
The result shows in [configs](https://github.com/h1063135843/mmsegmentation/tree/segmenter_head/configs/segmenter#readme)

There is also [another implementation](https://github.com/EricKani/Segmenter-Based-on-OpenMMLab) of your segmenter.

Of course, it is the best way for yourself to open a new PR. I sincerely hope these two implementation may help you

**Comment by rstrudel** [comment]
Thank you @h1063135843 , I did not know about the implementation by EricKani, this is really valuable. I will get in touch with him.

**Comment by Junjun2016** [comment]
Hi @rstrudel 
Looking forward to your PR!

**Comment by rstrudel** [comment]
@Junjun2016 @MengzhangLI I am writing a script to convert [ViT-AugReg](https://github.com/google-research/vision_transformer#available-vit-models) jax models to usable mmseg checkpoints. Once I have a ViT backbone, is there any script or way for me to check the backbone performance on ImageNet validation set ?

**Comment by Junjun2016** [comment]
> Nevermind, sorry for the request I found a way. Setting `strict=true` in `load_state_dict` and check manually that the squared matrices were not transposed did the trick to check what was missing and fix it. I will try to propose a new PR soon.

Besides, you should compare the forward results between your original model and mmseg model by feed a specific image or a dummy input (e.g. torch.ones(1, 3, 256, 256)).

**Comment by MengzhangLI** [comment]
Segmenter has been merged in MMSegmentation so I would close this pr. 

Anyway thanks for your warm-hearted pr! Feel free to contact us anytime you want!

**Comment by OpenMMLab-Coodinator** [comment]
Hi @h1063135843 ！We are grateful for your efforts in helping improve this open-source project during your personal time. 
Welcome to join OpenMMLab Special Interest Group (SIG) , where you can share your experiences, ideas, and build connections with like-minded peers. 

To join the SIG channel, simply message moderator— OpenMMLab on Discord or briefly share your open-source contributions in the #introductions channel and we will assist you. Look forward to seeing you there! Join us ：https://discord.gg/UjgXkPWNqA

If you have a WeChat account，welcome to join our community on WeChat. You can add our assistant ：openmmlabwx. Please add "mmsig + Github ID" as a remark when adding friends：）
Thank you again for your contribution❤
